### PR TITLE
Use libgvc to avoid having to include the command line tools

### PR DESCRIPTION
**Pull request by dschult**
The goal of this PR is to remove dependency on most of graphviz.
Then we could depend only on a few libraries like libcgraph, libcdt, libgvc and make wheels for easy installation.
It's possible that doing things within a library instead of via system calls is faster too, but I suspect not much.

I still need to clean up a bunch but I think it is all working.   
I left the old code in ```A.layout_command_line()``` and ```A.draw_command_line()```.

### CODE REVIEW

**Comment by dschult** [comment]
OK... I got the system dependent bugs worked through.
I have to say that using the dot representation of the graph to define equals and hash, etc
is difficult since the spacing depends on system and graphviz version. Even on the same system and graphviz, copying a graph can add an extra space in the dot file.

I changed hash and ```__eq__``` to look at the nodes, edges and attributes rather than the string.
This will be backwards incompatible -- but the old system was sufficiently broken I doubt too many people were using it.  Still, this might do better with a new primary number in the version: v2.0 instead of v1.7.

I also tweaked some of the string methods: there will be some different outcomes, though to the extent that the library creates the same results as the command line tools, they should be the same.

I'm still looking at tred, acyclic and unflatten to see if we can include those via a library.

**Comment by jarrodmillman** [comment]
We probably want to improve the test coverage for ``nx_agraph``:
https://codecov.io/gh/networkx/networkx/src/master/networkx/drawing/nx_agraph.py

**Comment by dschult** [comment]
@CristiFati and @mgorny     Do you have any thoughts on this move of pygraphviz from command line tools to Graphviz layout library interface?  If it works, it will make it much easier to maintain and get working with Windows and to some extent Mac OSs.

Also, we're looking at a fairly major change here.... Are there other aspects of pygraphviz that you'd like to see improved while we're digging around?

**Comment by dschult** [comment]
So far, I have tried to retain all previous functionality -- though to get the old ```A.draw()``` you have to use ```A.draw_command_line()```.  So, I think this is safe to merge even if we eventually pull the plug on it, we still have all the old stuff.  At some point though we may want to pull out some of the command-line tools -- to be determined...

**Comment by dschult** [comment]
Another question about interface design as we switch to using the libraries.
- In the old API we can use "args" when calling command-line functions. These are convenient ways to set the drawing parameters. That can also be done using the default attribute dict-like A.graph_attr, A.node_attr, A.edge_attr.  In the new interface there's a new method on AGraph to manually parse an input "args" and insert the values to the default attribute dict-likes.  This is different behavior in that the old "args" only impacted the command being run. It was a one-off setting of attributes.  The new "args" treatment makes those changes persistent, so unless you reset the values, using "args" once will impact all later drawing as well. I could easily make the impact of "args" last for one draw action by unsetting the default attributes.

Which is better? Having the "args" input act for only one call of the ```draw``` method? Or having it assign to default dicts so so possibly impact future ```draw``` calls as well?

**Comment by rkrishnasanka** [comment]
Oh please make this PR happen, the graphviz dependency is killing me on a couple of projects.

**Comment by jarrodmillman** [comment]
Sadly this PR will not happen soon.  We were told by the graphviz developers that "This is unlikely to be a response you want to hear, but my recommendation is to continue to call command line tools. There are numerous known unfixed out-of-bounds memory accesses (grep the issue tracker for segfault or anything filed by Google Autofuzz) and I suspect there are further latent issues like that fixed in !1489 (merged). Expecting Graphviz libraries to be memory safe or re-entrant is unwise. We are working on fixing the situation, but there is too much work and too few hands."

https://gitlab.com/graphviz/graphviz/-/issues/1808

**Comment by rossbar** [comment]
One option that's been discussed for moving forward with this is to add the library interface as an "experimental" feature. It seems the PR is mostly structured in this way already, e.g. there are two methods here `draw` and `draw_command_line`, which implement drawing via the library and command line (i.e. the way it's currently done) interfaces, respectively.

An idea for how to include the experimental library interface code would be to wrap it up in a method prepended with `_` so users know not to depend on it. For example, mapping the current method names (as of 34e8e9a) `draw` -> `_draw` and `draw_command_line` back to `draw`. At first glance, it seems that if we did this for the draw & layout methods of AGraph, that would be all that was necessary to expose the new implementations while maintaining backwards compatibility.

In the long term, the re-implementation of the graphviz wrapper using a library-based interface instead of a command-line interface makes a lot of sense. However, given the [recommendation from the graphviz devs referenced in the above comment](https://github.com/pygraphviz/pygraphviz/pull/244#issuecomment-722682898), I'm not sure this will have much of a positive impact for pygraphviz users *now*. Overall, I'd be +1 on exposing the new draw and layout implementations as `_draw` and `_layout` as I don't think there's any risk in doing so, and it might help the graphviz devs/anyone else interested in experimenting.

File: pygraphviz/agraph.py [hunk 1]
```diff
@@ -1563,14 +1682,35 @@ def draw(self, path=None, format=None, prog=None, args=""):
 
     # some private helper functions
 
+    def _manually_parse_args(self, args, format=None, prog=None):
```
[the remark this answers is not in the document]
Comment (dschult) [reply pygraphviz/agraph.py:1734]: I think I got this basic argument parsing code to work now. I have uncommented the places where this is used.   Thanks for flagging this.

**Comment by dschult** [comment]
I think this PR is ready now. Thanks @rossbar for getting the name switch figured out enough to get me restarted. I think I've got my local testbed up to speed (not for windows of course) and I turned on the arg processing to allow things like:
`G.draw("filename.png", prog='neato', args="-Nshape=box -Edir=forward -Ecolor=red")`

I have some ideas for the windows version lacking neato, etc. but that's a different PR...

**Pull request merged**